### DISCUSSION TITLE
Working on a particular ORM query regarding hierarchical tags

### DISCUSSION THREAD

**Discussion by bZichett** [simolus3/drift#2429] [Q&A] [answered]
I've been having trouble implementing the following query using drift ORM and figured I would reach out for help.

Four normalized tables are involved that model hierarchical tags, each tag with a many to many "Project" (using ProjectTagEdge).

`Tag` (id + superfluous columns)
`Project` (id + superfluous columns)
`ProjectTagEdge` (project_id, tag_id)
`TagTagEdge` (parent_id, child_id)

Find the count for the `Tag` roots of all `Project`s

More descriptive and about assertions: Return a count for every Project ... that totals the # of tags of that project ... with no parent that exists with the **same** project (_is a "root" for that project.)_  This could probably be even clearer, but represents the problem in steps better than the paragraph above.

But since each Tag can exist under different Projects, the query becomes a lot more complex - more complex than the scalar sub query example [here](https://drift.simonbinder.eu/docs/getting-started/expressions/#scalar-subqueries) which assumes a one to one relationship rather than many to many.

I am led to believe the solution will require a combination of `notExistsQuery` expressions and subquery expressions.  I have a test suite for test driven development modeling this complex case but after a lot of iterative exploration, I still can't seem to get this implementation right.  

Note: There is an extension to this implementation that will reuse the logic and returns those models rather than the count.

Perhaps after getting to the bottom of this, we could update the Drift documentation with more complex ORM examples.  Or if I missed something in the documentation, please link it.  Any help is much appreciated.

**Comment by simolus3** [comment]
It might just be too late, but I think I need some clarification on your query :D Tags belong to n projects and can have multiple parents as well?

Now, which tags are we counting? Those that directly belong to a project (have an entry in `ProjectTagEdge`) and don't have a parent that belongs to the same project? That would require a recursive lookup in `TagTagEdge`, right? (Since a "grandparent" tag could cause a tag to be excluded)?

**Reply by bZichett** [reply]
I figured I could be clearer - I dont think its a time of the day thing :+1: 

> Tags belong to n projects and can have multiple parents as well?

Yes, so more like a graph when the context arises - and it can in my experience.  And projects are like Labels (that are used for access privilege amongst other functionality in the software I'm developing.)  Pseudo-Hierarchical Tags (Graph) able to be assigned to different "Projects" for declarative use in _that_ context.  But all of this is more about the software and may be tertiary to the ORM solution :thinking: 

> Now, which tags are we counting? Those that directly belong to a project (have an entry in ProjectTagEdge) and don't have a parent that belongs to the same project?

Yep.  This is a more direct recast of the original ask. 

Pseudo Code
(ProjectTagEdge where project.id = [Project].id) & (Not Exists any(TagTagEdge.parent.projects contains [Project].id) )

>  That would require a recursive lookup in TagTagEdge, right? (Since a "grandparent" tag could cause a tag to be excluded)?

Recursive logic should not be needed but only as a design decision.  If grandparent has the project, but an intermediate doesn't (the parent,) then both "grandparent" tag and "child" tag would show up top level, in that project partition, by this implementation - although I consider that an edge case at the moment and can be ideally avoided by the end user.

**Reply by simolus3** [reply, marked as answer]
You can do the non-recursive case in drift thanks to its support for subqueries, but it doesn't exactly look great:

```dart
Future<int> tagsForProject(int projectId) {
  final count = countAll(); // translates to COUNT(*), we use a where clause

  // Use selectOnly since we only care about the columns we explicitly select,
  // in this case the count
  final query = selectOnly(projectTags)..addColumns([count]);

  // Filter tags that are in the current project
  query.where(projectTags.projectId.equals(projectId));

  // And where there is no parent tag in the same project
  final projectOfParent = alias(projectTags, 'parent_project');
  final parentTagQuery = selectOnly(parentTags)
    ..addColumns([Constant(1)]) // We don't need any real columns since this gets wrapped in an exists
    ..join([
      innerJoin(projectOfParent,
          projectOfParent.tagId.equalsExp(parentTags.parentId))
    ])
    ..where(projectTags.projectId.equals(projectId))
    ..where(parentTags.childId.equalsExp(projectTags.tagId));

  query.where(notExistsQuery(parentTagQuery));

  return query.map((row) => row.read(count)!).getSingle();
}
```

Essentially, we count the tags in a project (first condition). Then, we define another query that returns a row for each parent tag in the same project (using an alias over `ProjectTagEdge` to make it unambigious - we're also selecting from that table in the initial query). The second query gets wrapped in a `notExistsQuery` condition to be added as a where clause for the initial query.

**Reply by bZichett** [reply]
Much appreciated guidance, Simon.  I had to make just one critical adjustment to the above: In my use case (see below), the second `where` clause requires the alias  to work as expected due to the `select` on tags which uses the same table.  Returning a list of tags with a count instead of a single count for a project adds additional complexity.

Proposed solution above calculates # of children, so removing some code, here is the complete query for returning children of a single project with their count.  A bit verbose but working _exactly_ as intended.  From here, I might be missing some simplifications but can play around with it since I have a passing test suite on a complex graph of tags.

``` dart
Future<List<TagWithChildCount>> childTagsOfProject(Project project) async {
    final childCountAlias = alias(db.tagTagEdges, 'childCount');
    final nestedCountExpr = childCountAlias.parent_uuid.count();

    final query = select(db.tags).join([
      innerJoin(
        db.projectTagEdges,
        db.projectTagEdges.tag_uuid.equalsExp(db.tags.uuid),
      ),
      leftOuterJoin(
        childCountAlias,
        childCountAlias.parent_uuid.equalsExp(db.tags.uuid),
      ),
    ]);

    final projectOfParent = alias(db.projectTagEdges, 'projectOfParent');
    final parentTagQuery = selectOnly(db.tagTagEdges)
      ..addColumns([Constant(1)])
      ..join([
        innerJoin(projectOfParent,
            projectOfParent.tag_uuid.equalsExp(db.tagTagEdges.parent_uuid))
      ])
      ..where(projectOfParent.project_uuid.equals(project.uuid))
      ..where(db.tagTagEdges.child_uuid.equalsExp(db.projectTagEdges.tag_uuid));

    final whereClause = db.projectTagEdges.project_uuid.equals(project.uuid)
      & notExistsQuery(parentTagQuery);

     final result = await (query
      ..where(whereClause)
      ..addColumns([nestedCountExpr])
      ..groupBy([db.tags.uuid])
      ..orderBy([OrderingTerm(expression: db.tags.index_text)])
    ).get();

    return result.map((r) {
      final nestedCount = r.read(nestedCountExpr);
      final tag = r.readTable(db.tags);
      return TagWithChildCount(tag, nestedCount ?? 0);
    }).toList();
}
```
